src/contradictory_claims/models/sbert_models.py

```diff
@@ -0,0 +1,382 @@
+
+    train_data = SentencesDataset(nli_reader.get_examples(), model=model.embedding_model)
+    train_data.label_type = torch.long
+    # some bug in sentence_transformer library causes it to be identified as float by default
+    train_dataloader_embed = DataLoader(train_data, shuffle=True, batch_size=batch_size)
+    train_loss_embed = losses.SoftmaxLoss(
+        model=model.embedding_model,
+        sentence_embedding_dimension=model.embedding_model.get_sentence_embedding_dimension(),
+        num_labels=train_num_labels)
+
+    val_nli_reader = NLIDataReader(df_val)
+    dev_data = SentencesDataset(val_nli_reader.get_examples(), model=model.embedding_model)
+    dev_data.label_type = torch.long
+    evaluator = EmbeddingSimilarityEvaluator(sentences1=df_val["sentence1"].values,
+                                             sentences2=df_val["sentence2"].values,
+                                             scores=df_val["label"].values / 2.,
+                                             batch_size=batch_size)
+    warmup_steps = math.ceil(len(train_dataloader_embed) * epochs / batch_size * 0.1)  # 10% of train data for warm-up
+
+    # now to train the final layer
+    train_dataset = ClassifierDataset(df_train, tokenizer=tokenizer)
+    val_dataset = ClassifierDataset(df_val, tokenizer=tokenizer)
+    if enable_class_weights==False:
+        class_weights=None
+    else:
+        class_weights = train_dataset.class_weights()
+
+    train_dataloader = DataLoader(dataset=train_dataset,
+                                  batch_size=batch_size,
+                                  collate_fn=collate_fn,
+                                  shuffle=True)
+    val_dataloader = DataLoader(dataset=val_dataset,
+                                batch_size=1,
+                                collate_fn=collate_fn)
+
+    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+    criterion = nn.CrossEntropyLoss(weight=class_weights.to(device))
+    optimizer = optim.Adam(model.parameters(), lr=learning_rate)
+    model.to(device)
+    accuracy_stats = {"train": [],
+                      "val": [],
+                      }
+    loss_stats = {"train": [],
+                  "val": [],
+                  }
+
+    print("------TRAINING STARTS----------")  # noqa: T001
+    ## train embedding layer
+    unfreeze_layer(model.embedding_model)
+    model.embedding_model.fit(train_objectives=[(train_dataloader_embed, train_loss_embed)],
+                              evaluator=evaluator,
+                              epochs=1,
+                              evaluation_steps=1000,
+                              warmup_steps=warmup_steps,
+                              )  # train the Transformer layer
+    freeze_layer(model.embedding_model)
+
+
+    for e in range(epochs):
+        train_epoch_loss = 0
+        train_epoch_acc = 0
+        model.train()
+
+        for sentence1, sentence2, label in tqdm(train_dataloader):
```

Comment:
Why is this loop required? Isn't the model already being trained when model.fit() is called in line 173?

Reply:
this loop is training only the last layer, the earlier fit method fine tunes the SBERT architecture before this layer
SBERT has a custom training defined in their library itself, using the embeddings generated from their layer and adding an addition layer on top of that

Reply:
I'm not sure I understand....
I see that the model.fit() method calculates a softmax (defined by "train_loss_embed"), and the corresponding crossentropy loss is calculated and backpropagated in lines 585-588 of "SentenceTransformer.py"...I think that's the last layer?
What do you mean by the "last layer" that this loop is training?

Reply:
.fit() method  is fine tuning the SBERT model and we are passing information from this layer as input to classification layer
The loop trains this classification layer only